### ISSUE TITLE
RRULE UNTIL values must be specified in UTC when DTSTART is timezone-aware

### ISSUE THREAD

**Issue by squiggleuk** [niccokunzmann/python-recurring-ical-events#4]
Hello,

getting the following error when I try and use the .at or .between functions...

> Traceback (most recent call last):
>   File "2.py", line 22, in <module>
>     events = recurring_ical_events.of(cal).at(a_date)
>   File "/usr/local/lib/python3.7/site-packages/recurring_ical_events.py", line 133, in at
>     return self.between(datetime, datetime + self._DELTAS[len(date) - 3])
>   File "/usr/local/lib/python3.7/site-packages/recurring_ical_events.py", line 176, in between
>     rule.rrule(rrulestr(rule_string, dtstart=event_start))
>   File "/usr/local/lib/python3.7/site-packages/dateutil/rrule.py", line 1731, in __call__
>     return self._parse_rfc(s, **kwargs)
>   File "/usr/local/lib/python3.7/site-packages/dateutil/rrule.py", line 1653, in _parse_rfc
>     tzinfos=tzinfos)
>   File "/usr/local/lib/python3.7/site-packages/dateutil/rrule.py", line 1560, in _parse_rfc_rrule
>     return rrule(dtstart=dtstart, cache=cache, **rrkwargs)
>   File "/usr/local/lib/python3.7/site-packages/dateutil/rrule.py", line 470, in __init__
>     'RRULE UNTIL values must be specified in UTC when DTSTART '
> ValueError: RRULE UNTIL values must be specified in UTC when DTSTART is timezone-aware

Any ideas? I'm using the following values for the respective functions...

start_date = (2019, 6, 13, 12, 00 ,00 , 00)
end_date =   (2019, 6, 13)
a_date = (2019, 6, 13)

**Comment by squiggleuk** [comment]
Example ical....

>BEGIN:VEVENT
DTSTAMP:20190613T171521Z
DTSTART;VALUE=DATE:20190124
DTEND;VALUE=DATE:20190125
SUMMARY:WfH
X-CONFLUENCE-CUSTOM-TYPE-ID:6972027e-37ba-4a3e-82cb-5690b51c4ab4
CATEGORIES:WFH
SUBCALENDAR-ID:7bdacef4-2d77-432b-bd52-10dffc67d4fb
PARENT-CALENDAR-ID:5552044f-86e9-4d75-b038-9779bccf7a96
PARENT-CALENDAR-NAME:
SUBSCRIPTION-ID:
SUBCALENDAR-TZ-ID:GB
SUBCALENDAR-NAME:Calendar name here
EVENT-ID:75697
EVENT-ALLDAY:true
CUSTOM-EVENTTYPE-ID:aaaaaaaa-37ba-4a3e-82cb-5690b51c4ab4
UID:20190119T053217Z--1927336845@domain.com
DESCRIPTION:
ORGANIZER;X-CONFLUENCE-USER-KEY=aaaaaaaaaaaaa4eb01585ecded610030;CN=Fred Blo
 ggs;CUTYPE=INDIVIDUAL:mailto:person@domain.com
RRULE:FREQ=WEEKLY;INTERVAL=1;BYDAY=TH
CREATED:20190119T053217Z
LAST-MODIFIED:20190216T121411Z
ATTENDEE;X-CONFLUENCE-USER-KEY=ff808181582474eb01585ecded610030;CN=Fred Blo
 ggs;CUTYPE=INDIVIDUAL:mailto:person@domain.com
EXDATE;VALUE=DATE:20190718
EXDATE;VALUE=DATE:20190314
EXDATE;VALUE=DATE:20190307
EXDATE;VALUE=DATE:20190228
EXDATE;VALUE=DATE:20190221
SEQUENCE:6
X-CONFLUENCE-SUBCALENDAR-TYPE:custom
STATUS:CONFIRMED
END:VEVENT

**Comment by niccokunzmann** [comment]
Hi @squiggleuk, thanks for all the information. From this on I have a clear path of how to resolve this error the TDD way, so future updates do not break anything.

1. Create the file with the ICAL content from above in the [test/calendars](https://github.com/niccokunzmann/python-recurring-ical-events/tree/master/test/calendars) folder, e.g. `issue-4-timezone.ics`
2. Write two tests: one for at and one for between, 6 lines, which look like [this](https://github.com/niccokunzmann/python-recurring-ical-events/blob/master/test/test_single_events.py)

    ```
    def test_...(calendars):
        events = calendars.issue_4_timezone.between(...)
        assert ... events ...
    ```

From there, we can see how to fix this. How much time and energy are you willing to spend?

**Comment by niccokunzmann** [comment]
To motivate the above: The code you wrote looks just right to me. I do not know, why this error is created. Neither you not the calendar specify any time zone.

**Comment by niccokunzmann** [comment]
@squiggleuk, #5 aims at solving this.
I was not able to reproduce the error. However, I found another mistake I made and solved this.
I wonder how you would like to go about it from here.

**Comment by niccokunzmann** [comment]
I think, I know why I can not reproduce it: 
>RRULE UNTIL values must be specified in UTC when DTSTART is timezone-aware

No date is time-zone aware: neither in the example ICS file nor in the dates you pass as input.
Would you like to try to find out which date causes this error and get back?

**Comment by squiggleuk** [comment]
I will try and have a look when I next have some spare time!

**Comment by Sepulta** [comment]
I'm having the same issue here, it suddenly appeared a few days ago.

I tried to single down the event that caused it. I did that with the following piece of code:
```
calendar = icalendar.Calendar.from_ical(fix_e)
for event in calendar.walk('vevent'):
    try:
        rec = recurring_ical_events.of(event).between(start_date, end_date)
    except Exception as e:
        print(e)
        print(event)
```

The error message and the event : 

```
RRULE UNTIL values must be specified in UTC when DTSTART is timezone-aware

VEVENT({'DTSTART': <icalendar.prop.vDDDTypes object at 0x7f8bcdeeb518>, 'DTEND': <icalendar.prop.vDDDTypes object at 0x7f8bcdeeb470>, 'RRULE': vRecur({'FREQ': ['WEEKLY'], 'WKST': ['SU'], 'UNTIL': [datetime.date(2019, 10, 23)], 'BYDAY': ['TH']}), 'DTSTAMP': <icalendar.prop.vDDDTypes object at 0x7f8bcdeeb588>, 'ORGANIZER': vCalAddress(blabla), 'UID': vText(blabla), 'ATTENDEE': vCalAddress(blabla), 'CREATED': <icalendar.prop.vDDDTypes object at 0x7f8bcdeeb940>, 'DESCRIPTION': vText(blabla), 'LAST-MODIFIED': <icalendar.prop.vDDDTypes object at 0x7f8bcdeeb550>, 'LOCATION': vText('b'''), 'SEQUENCE': 0, 'STATUS': vText('b'CONFIRMED''), 'SUMMARY': vText(blabla), 'TRANSP': vText('b'OPAQUE'')})
```
Within the whole list of vevents that have a RRULE property containing an UNTIL value, it seems to be the only one having an UNTIL in the form of datetime.date instead of datetime.datetime. It also misses the tzinfo due to that.

**Comment by niccokunzmann** [comment]
@Sepulta Thanks for that description!

I thought that there could be that problem. However, I was not able to produce it and it never turned up for me.
https://github.com/niccokunzmann/python-recurring-ical-events/blob/38f53a00987d863bb526c489a2179f6809e843eb/recurring_ical_events.py#L196

As [written here](https://github.com/niccokunzmann/python-recurring-ical-events/issues/4#issuecomment-501813245), having an ical file with this event would help add it to the tests.
Can you do `print(event.to_ical())` to get this?

**Comment by Sepulta** [comment]
Hi @niccokunzmann,

The Ical is added below. One of the things that happened with this event is that the invited guest/person did not have a Gmail account. Doubt that this could have any effect on the UNTIL properties of this event, but better to mention peculiarities I guess.

```
b'BEGIN:VEVENT\r\nSUMMARY:blabla\r\nDTSTART;TZID=Europe/London:20190801T140000\r\nDTEND;TZID=Europe/London:20190801T150000\r\nDTSTAMP:20190801T083416Z\r\nUID:blabla\r\nSEQUENCE:0\r\nRRULE:FREQ=WEEKLY;UNTIL=20191023;BYDAY=TH;WKST=SU\r\nCREATED:20190729T105342Z\r\nDESCRIPTION:blabla\r\nLAST-MODIFIED:20190801T064315Z\r\nLOCATION:\r\nSTATUS:CONFIRMED\r\nTRANSP:OPAQUE\r\nEND:VEVENT\r\n'
```

Perhaps, if the UNTIL object is not a datetime, a default time + timezone (based on other properties of this event) could be added to change it into a datetime obj again.

**Comment by niccokunzmann** [comment]
@squiggleuk @Sepulta, thanks again. I am able to [reproduce the error](https://travis-ci.org/niccokunzmann/python-recurring-ical-events/jobs/566922123#L387), see #10. If any of you like, you can create a pull-request to that branch and start working on that error.

**Comment by weidenrinde** [comment]
I have  a probably related error. For a Ical with a **date** DTSTART and a weekly repetition RRULE, I get a "can't compare datetime.datetime to datetime.date".

I traced the error down to the following:
1. The between functions add the RRULE with `rule.rrule(rrulestr(rule_string, dtstart=event_start))`
2. Later event_start is added   with `rule.rdate(c_event_start)`
3. rrule translates the rule to  `DTSTART:20190823T000000, RRULE:FREQ=WEEKLY;UNTIL=20190830T000000;BYDAY=FR`, magically creating a datetime from the date.
4. rrule wants to compare the datetime from the rrule and the date from the rdate and crashes.

I think the problem is that rule expects rule.rdates to be **datetime** not **date**, see [rrule docu](https://www.pydoc.io/pypi/python-dateutil-2.7.3/autoapi/rrule/index.html?highlight=rdate#rrule.rruleset.rdate), and recurring-ical-event does deliver a **date**

**Comment by niccokunzmann** [comment]
Oh, yes. So, the challenge is the following:
Rrule can contain these:

- date
- datetime without timezone
- datetime with timezone

And they can be added in arbitrary order.

**Comment by niccokunzmann** [comment]
#10 should solve this.
Thanks @weidenrinde, @squiggleuk and @Sepulta for taking the time to advance on this matter.
this is included in v0.1.9b0.

**Comment by KoljaWindeler** [comment]
hi, first up: great work! i'm using your package in a home-assistent component and it helps a lot!!

i have a similar problem. turns out that the exchange server has in the rrule -> until a datetime.datetime but dtstart/dtend are datetime.date and that crashes 

more informations here https://github.com/KoljaWindeler/ics/issues/7#issuecomment-617961090 there is also a zip attached that produces the error consistently 2 posts above. the events.ics

thanks kolja
